frame/staking/src/lib.rs

```diff
@@ -939,6 +961,217 @@ where
+		ledger: &StakingLedger<T>,
+	) -> Option<PageIndex> {
+		if Self::is_non_paged_exposure(era, validator) {
+			return match ledger.legacy_claimed_rewards.binary_search(&era) {
+				// already claimed
+				Ok(_) => None,
+				// Non-paged exposure is considered as a single page
+				Err(_) => Some(0),
+			}
+		}
+
+		// Find next claimable page of paged exposure.
+		let page_count = Self::get_page_count(era, validator);
+		let all_claimable_pages: Vec<PageIndex> = (0..page_count).collect();
+		let claimed_pages = ClaimedRewards::<T>::get(era, validator);
+
+		all_claimable_pages.into_iter().find(|p| !claimed_pages.contains(p))
+	}
+
+	/// Checks if exposure is paged or not.
+	fn is_non_paged_exposure(era: EraIndex, validator: &T::AccountId) -> bool {
+		<ErasStakersClipped<T>>::contains_key(&era, validator)
+	}
+
+	/// Returns validator commission for this era and page.
+	pub(crate) fn get_validator_commission(
+		era: EraIndex,
+		validator_stash: &T::AccountId,
+	) -> Perbill {
+		<ErasValidatorPrefs<T>>::get(&era, validator_stash).commission
+	}
+
+	/// Creates an entry to track validator reward has been claimed for a given era and page.
+	/// Noop if already claimed.
+	pub(crate) fn set_rewards_as_claimed(
+		era: EraIndex,
+		validator: &T::AccountId,
+		page: PageIndex,
+	) {
+		let mut claimed_pages = ClaimedRewards::<T>::get(era, validator);
+
+		// this should never be called if the reward has already been claimed
+		if claimed_pages.contains(&page) {
+			defensive!("Trying to set an already claimed reward");
+			// nevertheless don't do anything since the page already exist in claimed rewards.
+			return
+		}
+
+		// add page to claimed entries
+		claimed_pages.push(page);
+		ClaimedRewards::<T>::insert(era, validator, claimed_pages);
+	}
+
+	/// Store exposure for elected validators at start of an era.
+	pub(crate) fn set_validator_exposure(
+		era: EraIndex,
+		validator: &T::AccountId,
+		exposure: Exposure<T::AccountId, BalanceOf<T>>,
+	) {
+		let page_size = T::MaxExposurePageSize::get().defensive_max(1);
+		let max_page_count = T::MaxExposurePageCount::get();
+
+		let nominator_count = exposure.others.len();
+		let required_page_count = nominator_count
+			.defensive_saturating_add(page_size as usize - 1) /
+			page_size as usize;
+
+		// clip nominators if it exceeds the maximum page count.
+		let exposure = if required_page_count as PageIndex > max_page_count {
+			// sort before clipping.
+			let mut exposure_clipped = exposure;
+			let clipped_max_len = max_page_count.saturating_mul(page_size);
+
+			exposure_clipped.others.sort_by(|a, b| b.value.cmp(&a.value));
+			exposure_clipped.others.truncate(clipped_max_len as usize);
+			exposure_clipped
+		} else {
+			exposure
+		};
+
+		let (exposure_overview, exposure_pages) = exposure.into_pages(page_size);
+
+		<ErasStakersOverview<T>>::insert(era, &validator, &exposure_overview);
+		exposure_pages.iter().enumerate().for_each(|(page, paged_exposure)| {
```

Comment:
```suggestion
		exposure_pages.iter().take(max_page_count).enumerate().for_each(|(page, paged_exposure)| {
```

Both of my suggestions here are super defensive, maybe you should double check `exposure_pages` instead.

Reply:
I added a defensive assert above on `exposure_pages`. This step does not need to be defensive since nothing will panic here.